### DISCUSSION TITLE
Full trace with library calls included

### DISCUSSION THREAD

**Discussion by fira959** [namhyung/uftrace#1972] [Q&A] [answered]
My intent is to collect all functions calls that are made by either my application or the library functions that my applications calls.
Running uftrace with --libname will get me only the library function that my application called, but I also want to know all function calls made by that library function itself. In other words, if my application calls strcmp@libc.so.6() and that libc function itself calles malloc@libc.so.6(); I want to log this as well.

I assume this is possible if I compile the library in question with the -pg flags as well, is that correct?

If so, the next question would be if and how anyone has managed to compile glibc accordingly and gotten the intended results?

**Comment by namhyung** [comment, marked as answer]
Hello, you can use `-l` or `--nest-libcall` option to trace library calls inside a library.

If you build a library with `-pg` you will be able to trace local functions too.  But unfortunately libc is special and fundamental to implement uftrace itself so you cannot trace inside of glibc.

**Reply by fira959** [reply]
Is there any way to trace internal libc function calls at all?

Would it be an option to use a different implementation of libc in the target application? Perhaps it is possible to use uftrace build with glibc to trace a program build with musl-libc?

**Reply by namhyung** [reply]
I'm not sure.. libc functions are called in a very low level and while I tried to avoid recursive calls to uftrace/libmcount there might be a subtle situation that can confuse it.  But I haven't tried such a complex configuration.  Please go ahead and share the result if you're brave enough.

**Reply by fira959** [reply]
So I am not quite sure I fully understand why tracing libc is an issue but I went ahead and simply statically compiled uftrace with a normal version of glibc to try and trace a target program linked to a version of glibc that is compiled with the -pg parameter.

However, after adding -pg to glibc I am getting strange errors like this:
```

gcc -pg -Wl,-O1 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now          -Wl,-z,pack-relative-relocs  -nostdlib -nostartfiles -r -o /home/fira/glibc/src/glibc-build/elf/librtld.map.o /home/fira/glibc/src/glibc-build/elf/librtld.mapT.o '-Wl,-(' /home/fira/glibc/src/glibc-build/elf/dl-allobjs.os /home/fira/glibc/src/glibc-build/libc_pic.a -lgcc '-Wl,-)' -Wl,-Map,/home/fira/glibc/src/glibc-build/elf/librtld.mapT
/usr/bin/ld: /home/fira/glibc/src/glibc-build/libc_pic.a(getcwd.os): in function `__GI___getcwd':
/usr/src/debug/glibc/glibc/io/../sysdeps/unix/sysv/linux/getcwd.c:49: multiple definition of `__getcwd'; /home/fira/glibc/src/glibc-build/elf/dl-allobjs.os:/usr/src/debug/glibc/glibc/elf/../sysdeps/unix/sysv/linux/getcwd.c:49: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:1337: /home/fira/glibc/src/glibc-build/elf/librtld.map] Error 1
make[2]: Leaving directory '/home/fira/glibc/src/glibc/elf'
make[1]: *** [Makefile:484: elf/subdir_lib] Error 2
make[1]: Leaving directory '/home/fira/glibc/src/glibc'
make: *** [Makefile:9: all] Error 2
```

This issue only appears when I add -pg to the CFLAGS and I don't see why or how to avoid it. 
Is there an inherent reason why this fails that you could explain?

**Reply by namhyung** [reply]
Sorry I don't have a good explanation for that.

**Reply by fira959** [reply]
I have tried this approach with glibc, musl-libc and diet-libc now and while glibc and dietlibc failed with duplicate errors like the one above, all reported an error about `undefined reference to `mcount'` despite adding -pg to CFLAGS, LDFLAGS and LDLIBS.

No other library I compiled with -pg had this issue so far, therefore I suspect this to be an issue with the added mcount code. Where is this mcount function defined anyway?
